core/new-gui/src/app/workspace/service/workflow-graph/model/shared-model-change-handler.ts

```diff
@@ -301,38 +303,70 @@ export class SharedModelChangeHandler {
-            if (addedInputPort) {
-              this.onOperatorPortAdded(
-                operatorID,
-                true,
-                (addedInputPort as Y.Map<any>[])[0].toJSON() as PortDescription
-              );
-            }
-            if (event.delta[0]?.delete || event.delta[1]?.delete) {
-              this.onOperatorPortRemoved(operatorID, true);
-            }
-          } else if (event.path[event.path.length - 1] === "outputPorts") {
-            const addedOutputPort = event.delta[1]?.insert ? event.delta[1]?.insert : event.delta[0]?.insert;
-            if (addedOutputPort) {
-              this.onOperatorPortAdded(
-                operatorID,
-                false,
-                (addedOutputPort as Y.Map<any>[])[0].toJSON() as PortDescription
-              );
-            }
-            if (event.delta[0]?.delete || event.delta[1]?.delete) {
-              this.onOperatorPortRemoved(operatorID, false);
-            }
           } else if (event.path.includes("operatorProperties")) {
             this.onOperatorPropertyChanged(operatorID, event.transaction.local);
+          } else if (event.path.includes("inputPorts")) {
+            this.handlePortEvent(event, operatorID, true);
+          } else if (event.path.includes("outputPorts")) {
+            this.handlePortEvent(event, operatorID, false);
+          } else {
+            throw new Error(`undefined operation on shared type: .${event}`);
           }
         }
       });
     });
   }
 
+  /**
+   * Handles the additon, deletion and deeper changes to operator ports.
+   * @param event
+   * @param operatorID
+   * @param isInput Since input and output ports are separate properties, need to access them differently.
+   * @private
+   */
+  private handlePortEvent(event: Y.YEvent<Y.Map<any>>, operatorID: string, isInput: boolean) {
+    if (event.path.length === 2) {
+      // Port added or deleted
+      const addedPort = event.delta[1]?.insert ? event.delta[1]?.insert : event.delta[0]?.insert;
+      if (addedPort) {
+        this.onPortAdded(operatorID, isInput, (addedPort as Y.Map<any>[])[0].toJSON() as PortDescription);
+      }
+      if (event.delta[0]?.delete || event.delta[1]?.delete) {
+        this.onPortRemoved(operatorID, isInput);
+      }
+    } else {
+      const changedOperator = this.texeraGraph.getOperator(operatorID);
+      if (event.path.includes("displayName")) {
+        // Display name changed (via shared text editor)
+        const changedPort = isInput
+          ? changedOperator.inputPorts[event.path[2] as number]
+          : changedOperator.outputPorts[event.path[2] as number];
+        this.texeraGraph.portDisplayNameChangedSubject.next({
+          operatorID: operatorID,
+          portID: changedPort.portID,
+          newDisplayName: event.target.toJSON() as unknown as string,
+        });
+      } else if (event.path.length >= 3) {
```

Comment:
is it possible not to rely on the number of paths to determine if it is a property change? do we have a more explicit check?

Reply:
There is no "number of paths" here. For now, like with an operator's property, we can only infer what an event is by its path, namely the "depth" of the Y object being modified within the operator (since it is a nested object and only the operator object can be listened on). Here the port property is: `Operator->Input Ports->Port Property->Specific Property`. Usually this inference logic works fine.

Reply:
It still sounds fragile to me and may expose potential issues. Are we using this inference logic in other Yjs-related components as well?

Reply:
This part of the Yjs-related code is all written like this. We may need to do a refactoring in the future for all such code.